src/Aspire.Hosting/Publishing/ManifestPublishingContext.cs

```diff
@@ -375,4 +501,44 @@ private void WriteContainerMounts(ContainerResource container)
             }
         }
     }
+
+    /// <summary>
+    /// Ensures that any <see cref="IResource"/> instances referenced by <paramref name="value"/> are
+    /// written to the manifest.
+    /// </summary>
+    /// <param name="value">The object to check for references that may be resources that need to be written.</param>
+    public void TryAddDependentResources(object? value)
+    {
+        _currentDependencyStack.Push(value);
+
+        if (value is IResource resource)
+        {
+            // add the resource to the ReferencedResources for now. After the whole model is written,
+            // these will be written to the manifest
+            _referencedResources.TryAdd(resource.Name, resource);
+        }
+        else if (value is IValueWithReferences objectWithReferences)
+        {
+            foreach (var dependency in objectWithReferences.References)
+            {
+                if (!_currentDependencyStack.Contains(dependency))
+                {
+                    TryAddDependentResources(dependency);
+                }
+            }
+        }
+
+        _currentDependencyStack.Pop();
+    }
+
+    private async Task WriteReferencedResources(DistributedApplicationModel model)
+    {
+        foreach (var resource in _referencedResources.Values)
+        {
+            if (!model.Resources.Contains(resource))
+            {
+                await WriteResourceAsync(resource).ConfigureAwait(false);
+            }
+        }
```

Comment:
nit: It's more efficient to remove resources that are already in the model, then write the what's left over.